src/main/java/com/synopsys/integration/alert/workflow/processor/JobProcessor.java

```diff
@@ -0,0 +1,126 @@
+
+        return topicContentList;
+    }
+
+    private Collection<NotificationContent> filterNotifications(final ProviderDescriptor providerDescriptor, final CommonDistributionConfig jobConfiguration, final Collection<NotificationContent> notificationCollection) {
+        final Predicate<NotificationContent> providerFilter = (notificationContent) -> jobConfiguration.getProviderName().equals(notificationContent.getProvider());
+        final Collection<NotificationContent> providerNotifications = applyFilter(notificationCollection, providerFilter);
+        final Collection<NotificationContent> filteredNotificationList = notificationFilter.extractApplicableNotifications(providerDescriptor.getProviderContentTypes(), jobConfiguration, providerNotifications);
+        return filteredNotificationList;
+    }
+
+    private List<TopicContent> collectTopics(final CommonDistributionConfig jobConfiguration, final Collection<NotificationContent> notificationCollection) {
+        final Optional<ProviderDescriptor> providerDescriptor = providerDescriptors.parallelStream().filter(descriptor -> jobConfiguration.getProviderName().equals(descriptor.getName())).findFirst();
+
+        if (!providerDescriptor.isPresent()) {
+            return Collections.emptyList();
+        } else {
+            final Collection<NotificationContent> notificationsForJob = filterNotifications(providerDescriptor.get(), jobConfiguration, notificationCollection);
+            if (notificationsForJob.isEmpty()) {
+                return Collections.emptyList();
+            }
+
+            final FormatType formatType = FormatType.valueOf(jobConfiguration.getFormatType());
+            final Map<String, TopicCollector> collectorMap = new HashMap<>();
+
+            //TODO need better performance to map the notification type to the processor
+            final Set<TopicCollector> providerTopicCollectors = providerDescriptor.get().createTopicCollectors();
+            for (final TopicCollector collector : providerTopicCollectors) {
+                for (final String notificationType : collector.getSupportedNotificationTypes()) {
+                    collectorMap.put(notificationType, collector);
+                }
+            }
+
+            notificationsForJob.parallelStream()
+                .filter(notificationContent -> collectorMap.containsKey(notificationContent.getNotificationType()))
+                .forEach(notificationContent -> collectorMap.get(notificationContent.getNotificationType()).insert(notificationContent));
+
+            return collectorMap.values().parallelStream().flatMap(collector -> collector.collect(formatType).stream()).collect(Collectors.toList());
+        }
+    }
+
+    //TODO notificationFilter also has this should it be made common?
+    private <T> List<T> applyFilter(final Collection<T> notificationList, final Predicate<T> filter) {
```

Comment:
We might not need this method if we just operated on Streams rather than Collections.